### ISSUE TITLE
Wekan Docker image build failing on Docker Hub

### ISSUE THREAD

**Issue by Barthalion** [wekan/wekan#928]
v0.16 wasn't tagged on Docker Hub, so it's impossible to pull it this way.

**Comment by xet7** [comment]
I'm not able to add tags, because I don't yet have access to Docker Hub of Wekan. You can always get latest version using tag "latest" as in docker install instructions, it's automatically built after every commit to repo. I'll try to contact @mquandalle sometime. More tasks to @mquandalle is at Wekan roadmap board that is linked on readme.

**Comment by rubberduck203** [comment]
Is there any chance we can open this back up. It's important to me to be able to pin to a version and upgrade only after I've thoroughly tested it locally. 0.17 just released and now if I need to restart my app/server I'll get the latest version instead of 0.16.

**Comment by rubberduck203** [comment]
So, this can worked around if you already have a copy of the 0.16 image locally. 
Tag your copy of the image yourself with this command.

```bash
docker tag 82e663c127dc mquandalle/wekan:0.16 
```

I'm going to also create a tag for the newest version. 
I'll see if I can set up a mirror.

**Comment by rubberduck203** [comment]
Mirror up here.
https://hub.docker.com/r/rubberduck/wekan/

**Comment by xet7** [comment]
@rubberduck203 

I have already sent support request to Docker Hub about adding tags, and also have direct contact to @mquandalle that has full access. Please be patient, it's taken care of. No need to open issues. :)

**Comment by rubberduck203** [comment]
Thanks @xet, but it's not so much a patience thing as a "You just scared me." thing. It's the kind of behavior that means I can't recommend this fantastic project to my clients as an ultra low cost option. Reproducible deployments matter. I apologize if I panicked, but I have plans to migrate to a new server and this, well, *panicked* me. =)

Anyway, the mirror is up for anyone who wants to pin to 0.16.

**Comment by xet7** [comment]
@rubberduck203 

I have sent message to @mquandalle that he adds tags to Docker Hub.

I don't know why some panic here, I try to be as friendly and welcoming as possible :)

**Comment by rubberduck203** [comment]
And I didn't mean to be unfriendly. 
The panic is because without the tags, it is impossible to pin to a known stable version. I simply can't afford to risk automatically upgrading (without a way to go back) just because I bounced a container.

**Comment by xet7** [comment]
@rubberduck203 

I'm currently creating new automated build with all tags to:
https://hub.docker.com/r/wekanteam/wekan/

Once it has built all tags, I will change all urls to point to there.

**Comment by xet7** [comment]
@rubberduck203 

How did you add docker tags to docker hub? on wekanteam/wekan my v0.16 and v0.17 builds fail and yours work.

**Comment by xet7** [comment]
@stephenmoloney 

Can you look why Docker Hub buils give errors?

https://hub.docker.com/r/mquandalle/wekan/

https://hub.docker.com/r/wekanteam/wekan/

**Comment by stephenmoloney** [comment]
@xet7 
I'll have a look now, I haven't been following wekan lately.
+1 for the tags and using another repo at https://hub.docker.com/r/wekanteam/wekan/~/dockerfile/ until you get the tags.
Just a pity it wasn't been tagged all along, it might have been easier to track where things went wrong.

**Comment by rubberduck203** [comment]
@xet I ran `docker images` to get the sha for v0.16 then ran the tag command above. Then I pulled latest and tagged it. 

I looked at the latest CI log, but didn't see anything related to docker. Can you share what you're seeing?

**Comment by xet7** [comment]
@stephenmoloney 

Build fails on dockerhub. I have not tried to build locally yet. I use Xubuntu 16.04 64bit for development. Could it be debian issue?

@rubberduck203 

Do you somehow push dockerhub tags back to dockerhub ?

**Comment by stephenmoloney** [comment]
Ok, I've looked into it and here is my take on the issue:

### Error Message from the DockerHub

```
kadira:flow-router: updating npm dependencies -- page, qs...

[91m
/home/wekan/.meteor/packages/meteor-tool/.1.3.5_1.1sksnpz++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/meteor-promise/promise_server.js:165
[0m
[91m      throw error;
            ^
[0m
[91mError: EXDEV, cross-device link not permitted '/home/wekan/app/packages/kadira:flow-router/.npm/package'
```

- The [commit](https://github.com/wekan/wekan/commit/92a34ba6a762b24f52da24edd8de29bd8cc3182d) *may* have lead to the first instance of the error. But I'd imagine this is a required dependency for the project so this is fine. 

- The issue seems to be with aufs and docker. Read more [here about that](https://github.com/meteor/meteor/issues/7852).

### Options/Solutions:

- This [merge request into meteor](https://github.com/meteor/meteor/pull/8491) *might* fix the issue. You would have to use meteor version 1.4.4 + to try this fix and as far as I know `wekan` isn't ready to upgrade meteor yet? 

- Revert back to a different version of `kadira:flow-router` for now and test if that works.

BTW, I will be making a pull request unrelated to this issue shortly.

**Comment by stephenmoloney** [comment]
@xet7 
Actually come to think of it, It might not be a good idea in general to have dependencies on 
a github repository - doesn't feel right to me - especially if it is outside the `wekan` organisation on 
github.

While I don't understand what `kadira:flow-router` is being used for, is the official ***published*** version of this package not suitable ?

**Comment by stephenmoloney** [comment]
Also, I cannot reproduce this error on my local machine - so I can't test any potential solution.
You could try reverting this [commit](https://github.com/wekan/wekan/commit/92a34ba6a762b24f52da24edd8de29bd8cc3182d).

**Comment by ocdtrekkie** [comment]
@stephenmoloney Yes, the double slash bug is because of the published version of kadira:flow-router. The patched version is in the Wekan organization on GitHub.

**Comment by weaktyper** [comment]
@ocdtrekkie , ah yes, I see that now on [this commit](https://github.com/wekan/wekan/commit/2031d107c432d340442dc12560902a9c1bedd02c). I guess upgrading meteor path is the way to go so.

**Comment by Serubin** [comment]
The fix for this is adding a `source app.env` in the dockerfile
I'll create a pr for the fix. Someone will have to give it a test for me though.

**Comment by xet7** [comment]
@stephenmoloney 

When building Wekan I got this error:
```
/home/user/repos/meteor/.meteor/isopackets/ddp/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:165
      throw error;
            ^
Error: ENOSPC, write
./rebuild-wekan.sh: line 20: cd: .build/bundle/programs/server: File or directory does not exist
```

I'll try to figure out why this happens.

Previously I did change used Dockerfile npm version to the one I use, but I'm not sure are there differences when I build on Xubuntu 16.04 64bit and that Debian wheezy that Docker image is based on.

**Comment by xet7** [comment]
Could be related to:
https://github.com/onmodulus/docker-build-meteor/issues/29

**Comment by xet7** [comment]
Doh, that error in building Wekan was only because I had full harddisk :) 

Now Wekan builds and works fine :)

But I don't know about that Docker Hub error yet.

**Comment by stephenmoloney** [comment]
I think the best bet is upgrading meteor version to 1.4.4 + as first port of call.

**Comment by stephenmoloney** [comment]
For reference, work is underway at this PR https://github.com/wekan/wekan/pull/957 to try and update meteor.

We just need to work through the bugs there.

**Comment by stephenmoloney** [comment]
@xet7 
Can you rename this issue to `Wekan build failing`

**Comment by schneivo77** [comment]
So what is the status about this topic?
I'm currently using the docker image from the repository at https://hub.docker.com/r/mquandalle/wekan/ but according to the Build Details, the last successful build was 1 month ago.
Will this ever be fixed? Or should I change the docker repository to https://hub.docker.com/r/wekanteam/wekan/ ?
But there the latest version has the tag "meteor-1.4" and not the version number of the latest release. This confuses me.

**Comment by stephenmoloney** [comment]
Hi,

`the last successful build was 1 month ago. Will this ever be fixed?` 
- This is likely to be fixed in release of branch `meteor-1.4` once that has stabilized. I can't say when.

`Or should I change the docker repository to https://hub.docker.com/r/wekanteam/wekan/ ?`
- Yes, for now, you should change to that repository as tagged releases are not currently possible at
`https://hub.docker.com/r/mquandalle/wekan/` due to lack of access and we need tagged releases.

`But there the latest version has the tag "meteor-1.4" and not the version number of the latest release. ` 
- I think that the latest release does not build correctly on the dockerhub. So have to wait for next release before this is fixed. 

Long story short, you can use `meteor-1.4` on `https://hub.docker.com/r/wekanteam/wekan/builds/`

**Comment by schneivo77** [comment]
OK, thank you! :+1:

**Comment by xet7** [comment]
Although some contributors have made all of meteor-1.4 branch running on Docker possible, I have not been able to replicate meteor-1.4 branch working outside Docker yet with installing from source, as seen at:
https://github.com/wekan/wekan/pull/1010#issuecomment-297735373

When some contributor helps me with that and documenting those steps to wiki, it's much easier for me to continue testing and accepting pull requests, and move meteor-1.4 branch to devel and master branches, so I can then continue making new releases.

**Comment by xet7** [comment]
- I built meteor-1.4 branch and merged it to devel and master
- I merged some pull requests, and translation updates.
- I updated install from source instructions:
https://github.com/wekan/wekan/wiki/Install-and-Update#install-manually-from-source

**Comment by xet7** [comment]
Today I have released Wekan v0.19, it's been build successfully and tagged at:
https://hub.docker.com/r/wekanteam/wekan/tags/

Some other tags have also been successfully built at:
https://quay.io/repository/wekan/wekan?tab=tags

For example:

```
docker run -d --restart=always --name wekan-db mongo:3.2.13

docker run -d --restart=always --name wekan --link "wekan-db:db" -e "MONGO_URL=mongodb://db" -e "ROOT_URL=http://localhost:8080" -p 8080:80 quay.io/wekan/wekan:v0.19
```
